vicon_transformer/transform.py

```diff
@@ -0,0 +1,74 @@
+from __future__ import annotations
+
+import typing
+
+import numpy as np
+import numpy.typing as npt
+from scipy.spatial.transform import Rotation
+
+
+class Transformation:
+    """Represents a 3d-transformation consisting of rotation and translation."""
+
+    def __init__(
+        self,
+        rotation: typing.Optional[typing.Union[Rotation, npt.ArrayLike]] = None,
+        translation: typing.Optional[npt.ArrayLike] = None,
+    ) -> None:
+        """
+        Args:
+            rotation: The rotation part of the transformation either as a
+                :class:`scipy.spatial.transform.Rotation` instance or a quaternion in
+                (x, y, z, w) format.  If not set, no rotation is done.
+            translation: The translation part of the transformation.  If not set, no
+                translation is done.
+        """
+        if rotation is None:
+            self.rotation = Rotation.identity()
+        elif isinstance(rotation, Rotation):
+            self.rotation = rotation
+        else:
+            # assume rotation is given as quaternion
+            self.rotation = Rotation.from_quat(rotation)
+
+        if translation is None:
+            self.translation = np.zeros(3)
+        else:
+            self.translation = np.asarray(translation)
+
+    @classmethod
+    def identity(cls) -> Transformation:
+        """Create a indentity transformation.
+
+        This is equivalent to using the constructor with default arguments but exists
+        for clarity and consistency with the Rotation class.
+        """
+        return cls()
```

Comment:
maybe the default arguments values for the constructor should be the one of identities (instead of None). Otherwise it may be easy to modify the constructor, and forget that by default it should set the identity for the translation and the rotation.

Reply:
This is problematic with mutable types like arrays.  The default value is a static value instantiated only once (probably when the module is loaded).  So when the value is changed, this affects the default value of future calls.  To illustrate:
```python
def foo(x=[0, 0, 0]):
    x[0] += 1
    print(x)

foo()  # prints '[1, 0, 0]'
foo()  # prints '[2, 0, 0]'
foo()  # prints '[3, 0, 0]'
```
Using `None` and then creating the default value in the constructor is a bit ugly but avoids this issue.

Reply:
Btw: I only learned about this issue recently through bugbear (a flake8 plugin that adds more checks), so I can highly recommend that :)